libcodechecker/analyze/analysis_manager.py

```diff
@@ -250,8 +250,31 @@ def check(check_data):
             if os.path.exists(rh.analyzer_result_file):
                 reanalyzed = True
 
-            # Fills up the result handler with the analyzer information.
-            source_analyzer.analyze(rh, analyzer_environment)
+            # Need to capture the "function pointer" returned by
+            # setup_process_timeout as reference, so that we may call it later.
+            # (By default, let's say that the process finished gracefully.)
+            timeout_cleanup = [lambda _: False]
```

Comment:
Any reason for this being a list?

Reply:
The comment explains it right above.

Reply:
`setup_process_timeout` returns a function pointer, but the function which calls `setup_...` is called through a callback. We need to capture this function pointer so we may call it later when the analyzer finishes (a few lines after this.)

Reply:
If we'd simply say `timeout_cleanup = setup_...` in the embedded function callback of `run_analyzer`, that assignment creates a local scope and shadows this variable, there is no way to retrieve the actual function pointer.

But lists and dicts are captured by reference, so we can assign to it and still use the value later on. By the time `run_analyzer()` returns, the analyzer has finished: either via exiting by itself, or the watcher killing it. So the callback has executed (because it executes at the **creation** of the subprocess), which means this function points to an actual closure of `setup_process_timeout.__cleanup_timeout()`. Which we can use to figure out if the exit of the analyzer (and return from `run_analyzer()`) happened from the analyzer quitting, or we murdering it.

Reply:
(Okay, I put a better explanation about this into the code too.)